### ISSUE TITLE
About yii2-mongodb find()->one

### ISSUE THREAD

**Issue by netyum** [yiisoft/yii2#6506]
When I use the mongo-php-driver 1.5.x version, it's ok, but an error when using the 1.6.x! I find the reason
cursor->hasNext() has a value, but $cursor->getNext() no value, therefore, is not 1.6.x tonal syntax, or a bug, but I did not find to submit issue for mongo-php-driver . Please follow.

**Comment by im-kulikov** [comment]
this problem not resolved?
Dotdeb now has php5-mongo_1%3a1.6.0-1~dotdeb.2_amd64.deb for Debian 7

Stack trace:
```php
PHP Warning 'yii\base\ErrorException' with message 'Invalid argument supplied for foreach()' 

in /project/vendor/composer/yiisoft/yii2/db/BaseActiveRecord.php:1047

Stack trace:
#0 /project/vendor/composer/yiisoft/yii2/db/BaseActiveRecord.php(1047): yii\base\ErrorHandler->handleError(2, 'Invalid argumen...', '/project...', 1047, Array)
#1 /project/vendor/composer/yiisoft/yii2/db/ActiveQueryTrait.php(132): yii\db\BaseActiveRecord::populateRecord(Object(app\models\Setting), NULL)
#2 /project/vendor/composer/yiisoft/yii2-mongodb/ActiveQuery.php(215): yii\mongodb\ActiveQuery->createModels(Array)
#3 /project/vendor/composer/yiisoft/yii2-mongodb/ActiveQuery.php(154): yii\mongodb\ActiveQuery->populate(Array)
#4 /project/app/components/Settings.php(31): yii\mongodb\ActiveQuery->one()
#5 /project/vendor/composer/yiisoft/yii2/base/Object.php(107): app\components\Settings->init()
#6 [internal function]: yii\base\Object->__construct(Array)
#7 /project/vendor/composer/yiisoft/yii2/di/Container.php(362): ReflectionClass->newInstanceArgs(Array)
#8 /project/vendor/composer/yiisoft/yii2/di/Container.php(147): yii\di\Container->build('app\\components\\...', Array, Array)
#9 /project/vendor/composer/yiisoft/yii2/BaseYii.php(344): yii\di\Container->get('app\\components\\...', Array, Array)
#10 /project/vendor/composer/yiisoft/yii2/di/ServiceLocator.php(133): yii\BaseYii::createObject(Array)
#11 /project/vendor/composer/yiisoft/yii2/base/Application.php(302): yii\di\ServiceLocator->get('settings')
#12 /project/vendor/composer/yiisoft/yii2/web/Application.php(63): yii\base\Application->bootstrap()
#13 /project/vendor/composer/yiisoft/yii2/base/Application.php(267): yii\web\Application->bootstrap()
#14 /project/vendor/composer/yiisoft/yii2/base/Object.php(107): yii\base\Application->init()
#15 /project/vendor/composer/yiisoft/yii2/base/Application.php(206): yii\base\Object->__construct(Array)
#16 /project/web/index.php(40): yii\base\Application->__construct(Array)
#17 {main}
```

I found the following code in `mongo-php-driver` (maybe it will help):

`v1.5.8` - https://github.com/mongodb/mongo-php-driver/blob/1.5.8/cursor.c#L234:

```c
	if (!cursor->started_iterating) {
		MONGO_METHOD(MongoCursor, doQuery, return_value, getThis());
		cursor->started_iterating = 1;
	}
```

`v1.6.0` - https://github.com/mongodb/mongo-php-driver/blob/1.6.0/cursor.c#L240:

```c
	if (!cursor->started_iterating) {
		php_mongo_runquery(cursor TSRMLS_CC);
		if (EG(exception)) {
			RETURN_NULL();
		}
		cursor->started_iterating = 1;
	}
```
Now, in Yii (don't pretend to be correct, but in that case - it works ;) ):

https://github.com/yiisoft/yii2/blob/master/extensions/mongodb/Query.php#L147-L172
```php
    protected function fetchRowsInternal($cursor, $all, $indexBy)
    {
        $result = [];
        if ($all) {
            foreach ($cursor as $row) {
                if ($indexBy !== null) {
                    if (is_string($indexBy)) {
                        $key = $row[$indexBy];
                    } else {
                        $key = call_user_func($indexBy, $row);
                    }
                    $result[$key] = $row;
                } else {
                    $result[] = $row;
                }
            }
        } else {
            if ($cursor->hasNext()) {
                $result = $cursor->getNext();
            } else {
                $result = false;
            }
        }
        return $result;
    }
```

I replace with:

```php
    protected function fetchRowsInternal($cursor, $all, $indexBy)
    {
        $result = [];

        foreach ($cursor as $row) {
            // if we need all - no changes:
            if ($all) {
                if ($indexBy !== null) {
                    if (is_string($indexBy)) {
                        $key = $row[$indexBy];
                    } else {
                        $key = call_user_func($indexBy, $row);
                    }
                    $result[$key] = $row;
                } else {
                    $result[] = $row;
                }
            // otherwise get one and break:
            } else {
                $result = $row;

                break;
            }
        }

        return $result;
    }
```

**Comment by profak** [comment]
@im-kulikov, thanks - solutions works. Problem still not resolved.

**Comment by im-kulikov** [comment]
@profak @cebe @netyum  if need, i can make pull request =)

**Comment by cebe** [comment]
@im-kulikov of course, thanks in advance :)

**Comment by im-kulikov** [comment]
@cebe yep, create #7141

**Comment by im-kulikov** [comment]
small fixes for pull request:

```php
    protected function fetchRowsInternal($cursor, $all, $indexBy)
    {
        $result = $all ? [] : false;

        foreach ($cursor as $row) {
            // if we need all - no changes:
            if ($all) {
                if ($indexBy !== null) {
                    if (is_string($indexBy)) {
                        $key = $row[$indexBy];
                    } else {
                        $key = call_user_func($indexBy, $row);
                    }
                    $result[$key] = $row;
                } else {
                    $result[] = $row;
                }
            // otherwise get one and break:
            } else {
                $result = $row;

                break;
            }
        }

        return $result;
    }
```

Otherwise, if the record is not found with a single search - returns an empty object.

**Comment by debasishgt** [comment]
I am new to yii so please pardon if i am making any silly mistake. I am using yi2 Adv template and the mongodb extensions. Without the above changes my registration works fine but login fails with the error "Invalid argument supplied for foreach()". With these changes the login works fine but registration fails (validation error) saying username already present. Any pointers?

**Comment by im-kulikov** [comment]
@debasishgt is you try changes in my comment (https://github.com/yiisoft/yii2/issues/6506#issuecomment-72552988)?

**Comment by debasishgt** [comment]
I tried :+1: protected function fetchRowsInternal($cursor, $all, $indexBy)
     {
         $result = [];
-        if ($all) {
-            foreach ($cursor as $row) {
+
+        foreach ($cursor as $row) {
+            if ($all) {
                 if ($indexBy !== null) {
                     if (is_string($indexBy)) {
                         $key = $row[$indexBy];
 @@ -159,12 +160,10 @@ protected function fetchRowsInternal($cursor, $all, $indexBy)
                 } else {
                     $result[] = $row;
                 }
-            }
-        } else {
-            if ($cursor->hasNext()) {
-                $result = $cursor->getNext();
             } else {
-                $result = false;
+                $result = $row;
+
+                break;
             }
         }

**Comment by im-kulikov** [comment]
@debasishgt `Without the above changes my registration works fine but login fails with the error "Invalid argument supplied for foreach()".` says that you use latest version of mongo-php-driver.

When I wrote the first to comment with my decision, I forgot to check with the help of test code for my tasks solution approach.

Later, when I finished my project testing - found the problem. The problem was in the foregoing comments.

If MongoCursor return null, but as a result has already been recorded array - the result of a method of `ActiveRecord::find()->one()` was an instance of an empty object.

Try next solution:

```php
    protected function fetchRowsInternal($cursor, $all, $indexBy)
    {
        $result = $all ? [] : false;

        foreach ($cursor as $row) {
            // if we need all - no changes:
            if ($all) {
                if ($indexBy !== null) {
                    if (is_string($indexBy)) {
                        $key = $row[$indexBy];
                    } else {
                        $key = call_user_func($indexBy, $row);
                    }
                    $result[$key] = $row;
                } else {
                    $result[] = $row;
                }
            // otherwise get one and break:
            } else {
                $result = $row;

                break;
            }
        }

        return $result;
    }
```

**Comment by im-kulikov** [comment]
For not to be misunderstandings, I will fix my first comment to no one more is no problem with the wrong solution to this problem

**Comment by debasishgt** [comment]
I get this with the above code:
MongoDB Exception – yii\mongodb\Exception
localhost:27017: Can't canonicalize query: BadValue unknown operator: $id ↵
Caused by: MongoCursorException
localhost:27017: Can't canonicalize query: BadValue unknown operator: $id
in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\Query.php at line 151

**Comment by im-kulikov** [comment]
@debasishgt can you take stack trace of problem?

**Comment by im-kulikov** [comment]
@debasishgt If I correctly understand, an Exception appears at this point (https://github.com/yiisoft/yii2/pull/7141/files#diff-ead70be42d9441afe990e3325636bc03R151)?

**Comment by debasishgt** [comment]
Exception
MongoDB Exception – yii\mongodb\Exception

localhost:27017: query failure
↵
Caused by: MongoCursorException

localhost:27017: query failure

in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\Query.php at line 165
1. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\Query.php at line 136
2. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\Query.php – yii\mongodb\Query::fetchRows(MongoCursor, false) at line 198
3. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\ActiveQuery.php – yii\mongodb\Query::one(null) at line 152
4. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\db\BaseActiveRecord.php – yii\mongodb\ActiveQuery::one() at line 101
5. in C:\xampp\htdocs\yii2mongodb\common\models\User.php – yii\db\BaseActiveRecord::findOne(['id' => ['$id' => '54d168fec0305d6409000029'], 'status' => 10]) at line 71
6. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\web\User.php – common\models\User::findIdentity(['$id' => '54d168fec0305d6409000029']) at line 295
7. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\web\User.php – yii\web\User::loginByCookie() at line 631
8. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\web\User.php – yii\web\User::renewAuthStatus() at line 176
9. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\web\User.php – yii\web\User::getIdentity() at line 346
10. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Component.php – yii\web\User::getIsGuest() at line 130
11. in C:\xampp\htdocs\yii2mongodb\frontend\views\layouts\main.php – yii\base\Component::__get('isGuest') at line 40
12. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\View.php – require('C:\xampp\htdocs\yii2mongodb\fron...') at line 325
13. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\View.php – yii\base\View::renderPhpFile('C:\xampp\htdocs\yii2mongodb\fron...', ['content' => '<div class="site-index"> <d...']) at line 247
14. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Controller.php – yii\base\View::renderFile('C:\xampp\htdocs\yii2mongodb\fron...', ['content' => '<div class="site-index"> <d...'], frontend\controllers\SiteController) at line 382
15. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Controller.php – yii\base\Controller::renderContent('<div class="site-index"> <d...') at line 368
16. in C:\xampp\htdocs\yii2mongodb\frontend\controllers\SiteController.php – yii\base\Controller::render('index') at line 70
17. frontend\controllers\SiteController::actionIndex()
18. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\InlineAction.php – call_user_func_array([frontend\controllers\SiteController, 'actionIndex'], []) at line 55
19. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Controller.php – yii\base\InlineAction::runWithParams([]) at line 151
20. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Module.php – yii\base\Controller::runAction('', []) at line 455
21. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\web\Application.php – yii\base\Module::runAction('', []) at line 83
22. in C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2\base\Application.php – yii\web\Application::handleRequest(yii\web\Request) at line 375
23. in C:\xampp\htdocs\yii2mongodb\frontend\web\index.php – yii\base\Application::run() at line 18
12131415161718    require(__DIR__ . '/../../common/config/main-local.php'),
    require(__DIR__ . '/../config/main.php'),
    require(__DIR__ . '/../config/main-local.php')
);
 
$application = new yii\web\Application($config);
$application->run();
 $_SERVER = [
    'MIBDIRS' => 'C:/xampp/php/extras/mibs',
    'MYSQL_HOME' => '\\xampp\\mysql\\bin',
    'OPENSSL_CONF' => 'C:/xampp/apache/bin/openssl.cnf',
    'PHP_PEAR_SYSCONF_DIR' => '\\xampp\\php',
    'PHPRC' => '\\xampp\\php',
    'TMP' => '\\xampp\\tmp',
    'HTTP_HOST' => 'localhost',
    'HTTP_CONNECTION' => 'keep-alive',
    'HTTP_CACHE_CONTROL' => 'max-age=0',
    'HTTP_ACCEPT' => 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'HTTP_USER_AGENT' => 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36',
    'HTTP_ACCEPT_ENCODING' => 'gzip, deflate',
    'HTTP_ACCEPT_LANGUAGE' => 'en-US,en;q=0.8',
    'HTTP_COOKIE' => '_identity=259c9970830a929c0f7a54381df284a27ee040a0a96302924f745b87ded1c57ba%3A2%3A%7Bi%3A0%3Bs%3A9%3A%22_identity%22%3Bi%3A1%3Bs%3A79%3A%22%5B%7B%22%24id%22%3A%2254d168fec0305d6409000029%22%7D%2C%22IiGomkgXlRVo2FynxTkSn1R4mujaoWTK%22%2C2592000%5D%22%3B%7D; _csrf=15b7f94cfd52bb561cdde179bdc5cd8b221da2dac2926255f740a3746130d5d6a%3A2%3A%7Bi%3A0%3Bs%3A5%3A%22_csrf%22%3Bi%3A1%3Bs%3A32%3A%225iyKObsgXboKg5iIODsFdqe1tLK8ng0r%22%3B%7D',
    'PATH' => 'C:\\ProgramData\\Oracle\\Java\\javapath;C:\\Program Files (x86)\\NVIDIA Corporation\\PhysX\\Common;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;C:\\Program Files (x86)\\ATI Technologies\\ATI.ACE\\Core-Static;C:\\xampp\\php;C:\\ProgramData\\ComposerSetup\\bin;C:\\mongodb\\bin',
    'SystemRoot' => 'C:\\Windows',
    'COMSPEC' => 'C:\\Windows\\system32\\cmd.exe',
    'PATHEXT' => '.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC',
    'WINDIR' => 'C:\\Windows',
    'SERVER_SIGNATURE' => '
Apache/2.4.10 (Win32) OpenSSL/1.0.1i PHP/5.6.3 Server at localhost Port 80

',
    'SERVER_SOFTWARE' => 'Apache/2.4.10 (Win32) OpenSSL/1.0.1i PHP/5.6.3',
    'SERVER_NAME' => 'localhost',
    'SERVER_ADDR' => '::1',
    'SERVER_PORT' => '80',
    'REMOTE_ADDR' => '::1',
    'DOCUMENT_ROOT' => 'C:/xampp/htdocs',
    'REQUEST_SCHEME' => 'http',
    'CONTEXT_PREFIX' => '',
    'CONTEXT_DOCUMENT_ROOT' => 'C:/xampp/htdocs',
    'SERVER_ADMIN' => 'postmaster@localhost',
    'SCRIPT_FILENAME' => 'C:/xampp/htdocs/yii2mongodb/frontend/web/index.php',
    'REMOTE_PORT' => '59516',
    'GATEWAY_INTERFACE' => 'CGI/1.1',
    'SERVER_PROTOCOL' => 'HTTP/1.1',
    'REQUEST_METHOD' => 'GET',
    'QUERY_STRING' => '',
    'REQUEST_URI' => '/yii2mongodb/frontend/web/index.php',
    'SCRIPT_NAME' => '/yii2mongodb/frontend/web/index.php',
    'PHP_SELF' => '/yii2mongodb/frontend/web/index.php',
    'REQUEST_TIME_FLOAT' => 1423012274.671000003814697265625,
    'REQUEST_TIME' => 1423012274,
];

$_COOKIE = [
    '_identity' => '259c9970830a929c0f7a54381df284a27ee040a0a96302924f745b87ded1c57ba:2:{i:0;s:9:"_identity";i:1;s:79:"[{"$id":"54d168fec0305d6409000029"},"IiGomkgXlRVo2FynxTkSn1R4mujaoWTK",2592000]";}',
    '_csrf' => '15b7f94cfd52bb561cdde179bdc5cd8b221da2dac2926255f740a3746130d5d6a:2:{i:0;s:5:"_csrf";i:1;s:32:"5iyKObsgXboKg5iIODsFdqe1tLK8ng0r";}',
];
Yii Framework
2015-02-04, 02:11:14

Apache/2.4.10 (Win32) OpenSSL/1.0.1i PHP/5.6.3
Yii Framework/2.0.3-dev
  Yii Debugger Yii 2.0.3-dev PHP 5.6.3 Status 500 Route site/index Log 39 1 Time 162 ms Memory 5.5 MB Asset Bundles 4 ›

**Comment by debasishgt** [comment]
Now the exception pops up when i am visiting index.php

**Comment by im-kulikov** [comment]
@debasishgt could you wrap your stack trace according to the Markdown syntax (https://guides.github.com/features/mastering-markdown/)?

**Comment by im-kulikov** [comment]
Please wrap `C:\xampp\htdocs\yii2mongodb\vendor\yiisoft\yii2-mongodb\Query.php` with markdown and add with stack trace. 

This is necessary in order to understand exactly where the error occurs.
So as of now, if I understand correctly, the problem occurs on line 165, which in my pull-request - is an empty string https://github.com/im-kulikov/yii2/blob/master/extensions/mongodb/Query.php#L165

You can also take a full file from my pull request and already using it to check how your project - https://github.com/im-kulikov/yii2/blob/master/extensions/mongodb/Query.php

**Comment by im-kulikov** [comment]
The problem, @debasishgt is that the query is constructed as follows:
```php
['id' => ['$id' => '<somevalue>'], ....]
```
But if I'm not mistaken, in Mongo used `_id` like a primary document key

**Comment by im-kulikov** [comment]
@debasishgt Now all works fine?

**Comment by im-kulikov** [comment]
If you use latest php-mongo-driver (this)[https://github.com/yiisoft/yii2/issues/6506#issuecomment-72780263] and (this)[https://github.com/yiisoft/yii2/issues/6506#issuecomment-72779996] not solve the problem.

For now, solution to resolve this problem, which appeared with the new php-mongo-driver is - https://github.com/im-kulikov/yii2/blob/master/extensions/mongodb/Query.php

**Comment by im-kulikov** [comment]
If I correctly understood from the source code the latest version of php-mongo-dr iver, the problem is that we are unable to now obtain result of a query using `$cursor->getNext ();`

**Comment by im-kulikov** [comment]
@debasishgt If the problem occurs in any case when you using  my decision, I propose to join forces and try to solve the problem together. Maybe something is going wrong in your case

**Comment by im-kulikov** [comment]
If you use [ActiveRecord with MongoDb](https://github.com/yiisoft/yii2/blob/master/extensions/mongodb/ActiveRecord.php) you need to rewrite [this](https://github.com/yiisoft/yii2-app-advanced/blob/master/common/models/User.php#L58-L64)

In mongodb hasn't `id` of document like it's in mysql or postrgresql.

Your query must be like this:

```php
    /**
     * @inheritdoc
     */
    public static function findIdentity($id)
    {
        return static::findOne(['_id' => new \MongoId($id), 'status' => self::STATUS_ACTIVE]);
    }
```

**Comment by Dotrox** [comment]
@im-kulikov, thank you! I am two days was trying to get this to work, changed name of parameter to '_id', but don't knew what to do with $id, it worked through once (I can't explain why).

I am tried your solution, but received ```Invalid object ID``` in this row when session expired.

My impression that Yii2 doesn't ready to work with MongoDB at this moment, too many things are not counted.

**Comment by im-kulikov** [comment]
@Dotrox you're welcome )

**Comment by debasishgt** [comment]
@im-kulikov
Sorry for the late response. I got busy with other things.
I tried your solution today.
My findIdentity($id) had 
"..findOne(['id' .." 
instead of 
"..findOne(['_id ..'"

Your solution fixed the issue. Thanks a lot for your help.

**Comment by im-kulikov** [comment]
@debasishgt you're welcome )

**Comment by im-kulikov** [comment]
current stable version of php-mongo-driver is `1.6.3`.
If they fix this bug, the pull-request can be closed.

https://github.com/mongodb/mongo-php-driver/pull/827 (v1.6.2)

**Comment by klimov-paul** [comment]
Issue resolved by commit 5d382ae
